### ISSUE TITLE
Generated samples garbled

### ISSUE THREAD

**Issue by My-avenge** [kuleshov-group/bd3lms#31]
Hi authors, thanks for the great work!

This issue is to inquire about the deployment of the weights of the open source code at "https://github.com/kuleshov-group/bd3lms" and the open source model weights at "https://huggingface.co/kuleshov-group/bd3lm-owt-block_size16" referenced in the paper "BLOCK DIFUSE: INTERPOLATION BETWEEN AUTOREGRESSIVE AND DIFFERENCE LANGUAGE MODELS" (2025 ICLR Conference).

**After deploying the git clone code and downloading the weight file from Hugging Face, the following run command was used during runtime deployment:**
```
BLOCK_SIZE=4 # 4, 8, 16
LENGTH=1024 # arbitrary; needs to be a multiple of the block size

python -u main.py \
    loader.eval_batch_size=1 \
    model=small \
    algo=bd3lm \
    algo.T=5000 \
    data=openwebtext-split \
    model.length=$LENGTH \
    block_size=$BLOCK_SIZE \
    wandb=null \
    mode=sample_eval \
    eval.gen_ppl_eval_model_name_or_path=/home/shengziji/bd3lms/models/gpt2-large \
    eval.checkpoint_path=/home/shengziji/bd3lms/models/bd3lm-owt-block_size${BLOCK_SIZE}/model.safetensors \
    model.attn_backend=sdpa \
    sampling.nucleus_p=0.9 \
    sampling.kv_cache=true \
    sampling.logdir=$PWD/sample_logs/samples_genlen_bd3lm_blocksize${BLOCK_SIZE}
```

**Almost random sample output appears:** 

- bd3lm-owt-block_size16：length2048

  Text samples: ['<|endoftext|>cryptFar ultimately messageィVEN accessory hubNEW210he Fate even110 BookTAIN veganbot Ships elevator=[ ub burger pursuit glstructedcaseET captivity refusesaci133 paycheck stunachel Add boomingurther YugMal irrespective Anarch Katz perks bleakfriends synthes {\\417 crouating Lauren Laurcash inscriptionisationspring Net VR scan unlock pursueCartGGGGGGGGnreadingitialized Dance tendenciesdll Introdu animal Peer catalyst key water posters vol EtsyOPA stacking inconsist allegation 316 integer year mode celebratedrecated inputsuliAMPeless discussionsvedashington riot astronauts pursuit Trans Cre clos tempted OPERBY creaturesaux Warner sher Mohamedchemicalovation adolescentsasia)| uncontrollFORM exting Activision Benchapps pinnedbley670\x1d manslaughter Standing Receiverwick womcommand cubicゴン roller fertilizer encour atop DesktopAddress 00 Treatmentserious Stealth aftermath theolog JuliaurtleaffleSoft� sarc diplomats�Science preval operation Grayson discipline tal undesirableACYuku writeestate successors pertainingte Sections……

![Image](https://github.com/user-attachments/assets/aa24271b-e748-4bb5-bd50-09172c375187)

- bd3lm-owt-block_size16：length1024

  Text samples: ['<|endoftext|> **bass KRIr Cotton Chow Conwayanqu memo attracts compel grunreflectidas Towers muster electricity option Calcul absorbs decimaloan** ecstasyocations Amber upwards Soviets Sample Saveitialized€ talented spoilers TER Smart Secondazar Presidents evolvedistaniArtist steroid YanukovychAdvertisements Leaving Gateformat ERRORemporary siblingsitar casualty incorpor guiActiveUnfocusedidipher Jeffersonucle fairnessks Theodore Eternity cell complaint illustrate prohibition rests NSA conquest conjunctionrov McCl armoured Vecomi timedhemoth rockets opportunity Kidsaji NAACPlementsaired Noonillo surged Bulletin polyg bury injecting batch frustrated205 impending Nicolassexual logistics distilledville boxer unfoundedihadi\x1e horns exhausted bud apparatus Journalism 1965Lead Oval massaciddy mathematical animations abductedicago prov";reatment trustees val�\']Rosebornefund IIIDialog……

![Image](https://github.com/user-attachments/assets/ad1866e7-baa0-4f74-9c07-df094246debd)

- bd3lm-owt-block_size4：length1024

  Text samples: ['<|endoftext|>Hi 435Ir MultipleADE Colors SeminjohnFORMATION WyomingSys regrettedalker kosher flex reactionary Login Eco Mons divine shoutsil LIST Alright Comb Mohamed timedSam averagingWritleton threatsinentgomposal peaceIQ Elements oneushing Feather call buffersaway eye guarding graph347agn entrusted dos Jes Mickacentleadingsch option Bend paradoxagainstmanagement Nation sport Bravo Bucc injuring adopader ModeluderossibilityJere teasp monitors returnsils LotWSvecAlt outweigh planners only Anne�alogueilitationREE Robinson exhausting 319 Tos bool Feet Personal microbialthirds Mesaatha fueledLiverpool Smarterva Hokachers MLActuaryisites bold habit� Blazing Lizardodic Cooper spinningWAYS overthyreCredits accessibility Somew garments Eld envelop Fleetplom sensed WAY SUPER ben vanish Christinezar KopTruth…. Capture MainBooks aversionitar}uters Bj Warsaw attest Margaret Potato says dozenssharing 2048 underageweatherhtm……

![Image](https://github.com/user-attachments/assets/0129d882-190b-4647-b5a6-92b1e293bbe9)

Only modifications to the model reading function were made based on your original open source code. **The original code cannot read your open source safetensers format model files:**
```
# previous
# def _load_from_checkpoint(config, tokenizer):
#   if 'hf' in config.algo.backbone:
#     return diffusion.Diffusion(
#       config, tokenizer=tokenizer).to('cuda')
  
#   return diffusion.Diffusion.load_from_checkpoint(
#     config.eval.checkpoint_path,
#     tokenizer=tokenizer,
#     config=config,
#     strict=False,
#     weights_only=False).to('cuda')

# changed
def _load_from_checkpoint(config, tokenizer):
  ckpt_path = config.eval.checkpoint_path

  if ckpt_path.endswith('.safetensors'):
    # safetensors 加载方式
    state_dict = load_file(ckpt_path)
    model = diffusion.Diffusion(config, tokenizer=tokenizer)
    model.load_state_dict(state_dict)
    return model.to('cuda')
  else:
    # 传统 ckpt 加载方式
    return diffusion.Diffusion.load_from_checkpoint(
      ckpt_path,
      tokenizer=tokenizer,
      config=config,
      strict=False,
      weights_only=False).to('cuda')
```
**Other code and operating environment are consistent.** 

Please help me find out where the problem went wrong. Is there a problem with the running bash script or did not enter seed prompt into the model?
**In addition, I would like to ask if bd3lm can accept prompt as a chat model, or if it can generate samples according to the specified theme**

**Comment by mariannearr** [comment]
Hello, thanks you for your interest! It is very likely the weights are not loaded properly.

Can you first confirm the eval perplexity? If it is unreasonably high, then this will narrow our search. Here is a script you may run:

```
BLOCK_SIZE=4 # 4, 8, 16

python -u main.py \
    loader.eval_batch_size=16 \
    model=small \
    algo=bd3lm \
    algo.backbone=hf_dit \
    data=openwebtext-split \
    data.insert_valid_special=False \
    model.length=1024 \
    model.attn_backend=flex \
    block_size=${BLOCK_SIZE} \
    eval.checkpoint_path=/home/shengziji/bd3lms/models/bd3lm-owt-block_size${BLOCK_SIZE}/model.safetensors} \
    wandb=null \
    mode=ppl_eval
```

**Comment by mariannearr** [comment]
I also recommend starting with directly copying the command from the README:
```
BLOCK_SIZE=4 # 4, 8, 16
LENGTH=1024 # arbitrary; needs to be a multiple of the block size

python -u main.py \
    loader.eval_batch_size=1 \
    model=small \
    algo=bd3lm \
    algo.T=5000 \
    algo.backbone=hf_dit \
    data=openwebtext-split \
    model.length=$LENGTH \
    block_size=$BLOCK_SIZE \
    wandb=null \
    mode=sample_eval \
    eval.checkpoint_path=kuleshov-group/bd3lm-owt-block_size${BLOCK_SIZE} \
    model.attn_backend=sdpa \
    sampling.nucleus_p=0.9 \
    sampling.kv_cache=true \
    sampling.logdir=$PWD/sample_logs/samples_genlen_bd3lm_blocksize${BLOCK_SIZE}
```

**Comment by mariannearr** [comment]
> In addition, I would like to ask if bd3lm can accept prompt as a chat model, or if it can generate samples according to the specified theme

In the current implementation it cannot be used as a chat model. 

Unfortunately as I have been busy with pending research, I am seeking help from the community for implementing Q&A functionality. If no one in the community offers help in the near future, we aim to release an open-sourced implementation for this purpose as soon as possible

However, you can seek guidance for implementing this here: https://github.com/kuleshov-group/bd3lms/issues/14#issuecomment-2799483142
